cmd/blsync/main.go

```diff
@@ -0,0 +1,243 @@
+	execSyncer := &execSyncer{
+		api:           beaconApi,
+		client:        makeRPCClient(ctx),
+		execRootCache: lru.NewCache[common.Hash, common.Hash](1000),
+	}
+	syncCommitteeTracker.SubscribeToNewHeads(execSyncer.newHead)
+	committeeSyncer.Start(syncCommitteeTracker)
+	syncCommitteeCheckpoint.TriggerFetch()
+	<-ctx.Done()
+	return nil
+}
+
+func callNewPayloadV1(client *rpc.Client, block *ctypes.Block) (string, error) {
+	var resp engine.PayloadStatusV1
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	err := client.CallContext(ctx, &resp, "engine_newPayloadV1", *engine.BlockToExecutableData(block))
+	cancel()
+	return resp.Status, err
+}
+
+func callForkchoiceUpdatedV1(client *rpc.Client, headHash, finalizedHash common.Hash) (string, error) {
+	var resp engine.ForkChoiceResponse
+	update := engine.ForkchoiceStateV1{
+		HeadBlockHash:      headHash,
+		SafeBlockHash:      finalizedHash,
+		FinalizedBlockHash: finalizedHash,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	err := client.CallContext(ctx, &resp, "engine_forkchoiceUpdatedV1", update, nil)
+	cancel()
+	return resp.PayloadStatus.Status, err
+}
+
+type execSyncer struct {
+	api           *api.BeaconLightApi
+	client        *rpc.Client
+	execRootCache *lru.Cache[common.Hash, common.Hash] // beacon block root -> execution block root
+}
+
+var statePaths = []string{
+	"[\"finalizedCheckpoint\",\"root\"]",
+	"[\"latestExecutionPayloadHeader\",\"blockHash\"]",
+}
+
+func (e *execSyncer) newHead(head types.Header) {
+	log.Info("Received new beacon head", "slot", head.Slot, "blockRoot", head.Hash())
+	if e.client == nil {
+		return
+	}
+	var (
+		blocks         []*ctypes.Block
+		finalizedRoots []common.Hash
+		parentFound    bool
+		maxBlockFetch  = 4
+	)
+	for {
+		proof, err := e.api.GetStateProof(head.StateRoot, statePaths, stateProofFormat)
+		if err != nil {
+			log.Error("Error fetching state proof from beacon API", "error", err)
+			break
+		}
+		var (
+			execBlockRoot       = common.Hash(proof.Values[execBlockIndex])
+			finalizedBeaconRoot = common.Hash(proof.Values[finalizedBlockIndex])
+			beaconRoot          = head.Hash()
+		)
+		block, err := e.api.GetExecutionPayload(beaconRoot, execBlockRoot)
+		if err != nil {
+			log.Error("Error fetching execution payload from beacon API", "error", err)
+			break
+		}
+		blocks = append(blocks, block)
+		finalizedRoots = append(finalizedRoots, finalizedBeaconRoot)
+		e.execRootCache.Add(beaconRoot, execBlockRoot)
+
+		if _, ok := e.execRootCache.Get(head.ParentRoot); ok {
+			parentFound = true
+			break
+		}
+		maxBlockFetch--
+		if maxBlockFetch == 0 {
+			break
+		}
+
+		if head, err = e.api.GetHeader(head.ParentRoot); err != nil {
+			log.Error("Error fetching header from beacon API", "error", err)
+			break
+		}
+	}
+	if blocks == nil {
+		return
+	}
+	if !parentFound && maxBlockFetch == 0 {
```

Comment:
Seems like the only purpose of this loop is to collect as many beacon-execution root as possible in the `e.execRootCache`. Is it really necessary? Personally I agree with martin that only the latest block is required for triggering sync. 

EL will handle reorg(in downloader) in case the provided new head header is not in the same chain as the previous local head.

Reply:
This second loop is not needed for triggering sync (indeed, for that the head block would be enough). Collecting `execRootCache` associations is a lot cheaper than fetching entire blocks and it is used to serve the finalized execution block correctly. From the beacon state I get the finalized consensus block root and also the current consensus block root -> exec block root association but the finalized root points to an older block, therefore I need to collect older consensus -> exec root associations in order to know the finalized exec block root (which is I believe a nice thing to have in Geth and it does not cost a lot of resources).

Reply:
I simplified this part as we discussed earlier. Now the beaconRoot -> execRoot associations are fetched by a separate function (so that it does not clutter up `newHead`). Also, due to popular demand, I removed the part that fetched multiple exec blocks and fed them into the engine API :) Now only the head block is fetched and the EL client can take care of the rest.